commit 615565a92519a83ff8433e7667c620505ba57f66
Author: Josh Cooper <josh@puppet.com>
Date:   Tue Jan 22 15:24:55 2019 -0800

    Raise EOF if response body content length is known

    If response body content length is known, then read the complete
    response body or raise EOF if it is truncated.

    [Bug #14972][ruby-core:88324]

diff --git a/lib/net/http/response.rb b/lib/net/http/response.rb
index 66132985d9..7c744d02f4 100644
--- a/lib/net/http/response.rb
+++ b/lib/net/http/response.rb
@@ -290,7 +290,7 @@ def read_body_0(dest)

       clen = content_length()
       if clen
-        @socket.read clen, dest, true   # ignore EOF
+        @socket.read clen, dest
         return
       end
       clen = range_length()
diff --git a/test/net/http/test_httpresponse.rb b/test/net/http/test_httpresponse.rb
index a03bb2e152..9d00276514 100644
--- a/test/net/http/test_httpresponse.rb
+++ b/test/net/http/test_httpresponse.rb
@@ -401,6 +401,21 @@
     end
   end

+  def test_raise_eof
+    io = dummy_io(<<EOS)
+HTTP/1.1 200
+Content-Length: 5
+Connection: close
+
+h
+EOS
+
+    res = Net::HTTPResponse.read_new(io)
+    assert_raise EOFError do
+      res.reading_body(io, true) {}
+    end
+
 private

   def dummy_io(str)
